server/app/views/components/FieldWithLabel.java

```diff
@@ -246,107 +243,183 @@ public FieldWithLabel showFieldErrors(boolean showFieldErrors) {
+    String[] referenceClasses =
+        referenceClassesBuilder.build().stream().map(ref -> ref + "-error").toArray(String[]::new);
+    return div(each(fieldErrors, error -> div(error.format(messages))))
+        .withId(id)
+        .withClasses(
+            StyleUtils.joinStyles(referenceClasses),
+            StyleUtils.joinStyles(BaseStyles.FORM_ERROR_TEXT_XS, Styles.P_1),
+            fieldErrors.isEmpty() || !showFieldErrors ? Styles.HIDDEN : "");
+  }
+
+  protected void genRandIdIfEmpty() {
     // In order for the labels to be associated with the fields (mandatory for screen readers)
     // we need an id.  Generate a reasonable one if none is provided.
     if (this.id.isEmpty()) {
       this.id = RandomStringUtils.randomAlphabetic(8);
     }
-    String fieldErrorsId = String.format("%s-errors", this.id);
-    if (fieldTag.getTagName().equals("textarea")) {
-      // Have to recreate the field here in case the value is modified.
-      ContainerTag textAreaTag = textarea().withType("text").withText(this.fieldValue);
-      fieldTag = textAreaTag;
-    } else if (this.fieldType.equals("number")) {
-      // Setting inputmode to decimal gives iOS users a more accessible keyboard
-      fieldTag.attr("inputmode", "decimal");
-
-      // Setting step to any disables the built-in HTML validation so we can use our
-      // custom validation message to enforce integers.
-      fieldTag.attr("step", "any");
-
-      // Set min and max values for client-side validation
-      if (this.minValue.isPresent()) {
-        fieldTag.attr("min", minValue.getAsLong());
-      }
-      if (this.maxValue.isPresent()) {
-        fieldTag.attr("max", maxValue.getAsLong());
-      }
-
-      // For number types, only set the value if it's present since there is no empty string
-      // equivalent for numbers.
-      if (this.fieldValueNumber.isPresent()) {
-        fieldTag.withValue(String.valueOf(this.fieldValueNumber.getAsLong()));
-      }
-    } else {
-      fieldTag.withValue(this.fieldValue);
+  }
+
+  protected LabelTag genLabelTag() {
+
+    return label()
+        .attr(Attr.FOR, this.id)
+        // If the text is screen-reader text, then we want the label to be screen-reader
+        // only.
+        .withClass(labelText.isEmpty() ? Styles.SR_ONLY : BaseStyles.INPUT_LABEL)
+        .withText(labelText.isEmpty() ? screenReaderText : labelText);
+  }
+
+  protected DivTag wrapInDivTag(Tag fieldTag, Tag labelTag, String fieldErrorsId) {
+    return div(
+            labelTag,
+            div(fieldTag, buildFieldErrorsTag(fieldErrorsId))
+                .withClasses(Styles.FLEX, Styles.FLEX_COL))
+        .withClasses(
+            StyleUtils.joinStyles(referenceClassesBuilder.build().toArray(new String[0])),
+            BaseStyles.FORM_FIELD_MARGIN_BOTTOM);
+  }
+
+  protected boolean getHasFieldErrors() {
+    return !fieldErrors.isEmpty() && showFieldErrors;
+  }
+
+  protected void numberTagApplyAttrs(Tag fieldTag) {
```

Comment:
Can this argument be specifies as an `InputTag`?

If so, we can call `.setStep` and `setValue` below rather than calling that attr setters directly.